ethdb/privateapi/ethbackend.go

```diff
@@ -147,3 +169,140 @@ func (s *EthBackendServer) Block(ctx context.Context, req *remote.BlockRequest)
+		return nil, err
+	}
+
+	if !isTrans {
+		// We are still syncing proof-of-work chain
+		return &remote.EngineExecutePayloadReply{
+			Status:          Syncing,
+			LatestValidHash: gointerfaces.ConvertHashToH256(currentHead),
+		}, nil
+	}
+	blockHash := gointerfaces.ConvertH256ToHash(req.BlockHash)
+
+	if s.numberSent > *headNumber {
+		// We are still syncing a commisioned payload
+		return &remote.EngineExecutePayloadReply{
+			Status:          Syncing,
+			LatestValidHash: gointerfaces.ConvertHashToH256(currentHead),
+		}, nil
+	}
+	// Let's check if we have parent hash, if we have it we can process the payload right now.
+	// If not, we need to commission it and reverse-download the chain.
+	var baseFee *big.Int
+	eip1559 := false
+
+	if req.BaseFeePerGas != nil {
+		baseFee = gointerfaces.ConvertH256ToUint256Int(req.BaseFeePerGas).ToBig()
+		eip1559 = true
+	}
+
+	// Decode transactions
+	reader := bytes.NewReader(nil)
+	stream := rlp.NewStream(reader, 0)
+	transactions := make([]types.Transaction, len(req.Transactions))
+
+	for i, encodedTransaction := range req.Transactions {
+		reader.Reset(encodedTransaction)
+		stream.Reset(reader, 0)
+		if transactions[i], err = types.DecodeTransaction(stream); err != nil {
+			return nil, err
+		}
+	}
+
+	// Extra data can go from 0 to 32 bytes, so it can be treated as an hash
+	var extra_data common.Hash = gointerfaces.ConvertH256ToHash(req.ExtraData)
+	header := types.Header{
+		ParentHash:  gointerfaces.ConvertH256ToHash(req.ParentHash),
+		Coinbase:    gointerfaces.ConvertH160toAddress(req.Coinbase),
+		Root:        gointerfaces.ConvertH256ToHash(req.StateRoot),
+		Bloom:       gointerfaces.ConvertH2048ToBloom(req.LogsBloom),
+		Random:      gointerfaces.ConvertH256ToHash(req.Random),
+		Eip3675:     true,
+		Eip1559:     eip1559,
+		BaseFee:     baseFee,
+		Extra:       extra_data.Bytes(),
+		Number:      big.NewInt(int64(req.BlockNumber)),
+		GasUsed:     req.GasUsed,
+		GasLimit:    req.GasLimit,
+		Time:        req.Timestamp,
+		MixDigest:   common.Hash{},
+		UncleHash:   types.EmptyUncleHash,
+		Difficulty:  serenity.SerenityDifficulty,
+		Nonce:       serenity.SerenityNonce,
+		ReceiptHash: gointerfaces.ConvertH256ToHash(req.ReceiptRoot),
+		TxHash:      types.DeriveSha(types.Transactions(transactions)),
+	}
+	// Our execution layer has some problems so we return invalid
+	if header.Hash() != blockHash {
+		return nil, fmt.Errorf("invalid hash for payload. got: %s, wanted: %s", common.Bytes2Hex(blockHash[:]), common.Bytes2Hex(header.Hash().Bytes()))
+	}
+	log.Info("Received Payload from beacon-chain", "hash", blockHash)
+	// Send the block over
+	s.numberSent = req.BlockNumber
+	s.reverseDownloadCh <- *types.NewBlock(&header, transactions, nil, nil)
+	// Check if current block is next for execution, if not, commission it and start
+	// Reverse-download the chain from its block number and hash.
+	if header.ParentHash != currentHead {
+		return &remote.EngineExecutePayloadReply{
```

Comment:
If you leave here, no-one will take the value off the `s.statusCh`, and it may be incorrectly read by the next invocation